### PR TITLE
Ehanc/#5 docker

### PR DESCRIPTION
**Pull request by cassjs**
Added Dockerfile and added chromedriver to .gitignore. 

Having issue with running program through CLI when I input _$ python3 vaccinewatcher/watcher.py_. 
`  File "vaccinewatcher/watcher.py", line 4, in <module>
    import elemental
ModuleNotFoundError: No module named 'elemental'`

When I use the Run and Debug mode in vscode, it runs fine. 

-----------------------
Edit: Removed original comment below. Added instructions to README.md file.

### CODE REVIEW

**Comment by Xuehong-pdx** [comment]
> Having issue with running program through CLI when I input _$ python3 vaccinewatcher/watcher.py_.
> ` File "vaccinewatcher/watcher.py", line 4, in <module> import elemental ModuleNotFoundError: No module named 'elemental'`
> 
> When I use the Run and Debug mode in vscode, it runs fine.
> 

I encountered an opposite problem.  Got the same error running from the Run and Debug mode in vscode, but no problem when I run  "$ python3 vaccinewatcher/watcher.py" from command line.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+
+WORKDIR /vaccinewatcher
+COPY . . 
+
+# Install dependencies
+COPY requirements.txt .
+RUN pip3 install -r requirements.txt 
+
+# Run application
+CMD [ "python", "vaccinewatcher/watcher.py" ]
```
Comment (lildonpancho) [inline Dockerfile:11]: This line should be `ENTRYPOINT` instead of `CMD` because this is a command-line script.

**Review by lildonpancho — CHANGES_REQUESTED** [review]

File: README.md [hunk 2]
```diff
@@ -40,16 +41,24 @@ brew install chromedriver
...
+```
 
-`vaccinewatcher` is on currently only available on Github since it's not stable yet:
+**Docker**\
+**What is Docker?** Docker is a platform for building, running, and shipping applications. Docker packages up an application with everything it needs and allows an app to run and function the same way on any user's local machine.
 
-```bash
-# from pypi
-pip install --upgrade vaccinewatcher
+**Download Docker Desktop** In order to run the Docker container, you’ll need to install Docker then run it on your local machine. There are versions available for Linux, Max, and Windows. Download the Docker desktop application: https://www.docker.com/get-started (If you’re using VScode, the docker extension can also be utilized, however, Docker Desktop will still need to run in the background on your local machine when running the program.
```
[the remark this answers is not in the document]
Comment (cassjs) [reply README.md]: Great input and feedback, thank you! After docker download, I added info about signup, then logging in from the terminal. I also added info about editing the application. Let me know if you have any feedback on that as well.

File: run.sh [hunk 3]
```diff
@@ -0,0 +1,5 @@
+# Build Docker image
+docker build -t vaccinewatcher . 
+
+# Run Docker Container
+docker run -i vaccinewatcher:latest
```
Comment (lildonpancho) [inline run.sh:7]: I would separate out the run and build commands into a run.sh and a build.sh.

File: run.sh [hunk 4]
```diff
@@ -0,0 +1,7 @@
+docker login
```
Comment (lildonpancho) [inline run.sh:1]: The user will not need to log in to Docker every time the script is run. I think that we should leave the instruction in the README on how to log in but take it out of the bash script.

**Review by lildonpancho — CHANGES_REQUESTED** [review]
Looks good! Nice job on the Docker stuff. A few more minor tweaks and it should be good to merge.

File: run.sh (hunk 3 above)
Comment (cassjs) [reply run.sh]: I separated these out and placed them in a scripts directory to make it cleaner.

File: run.sh (hunk 4 above)
Comment (cassjs) [reply run.sh]: Oh durr.🤪 Thank you for catching that!

File: scripts/run.sh [hunk 5]
```diff
@@ -0,0 +1,2 @@
+# Run Docker container
+docker run -i vaccinewatcher:latest
```
Comment (lildonpancho) [inline scripts/run.sh:2]: What is the `-i` here for?

File: README.md [hunk 6]
```diff
@@ -54,10 +54,18 @@ pip install chromedriver
...
+      $ docker login
+      
+* **Build Docker Image:** 
+      
+      $ bash scripts/build.sh
+      
+* **Run Application/Docker Container:**
       
-      $ bash run.sh
+      $ bash scripts/run.sh
```
Comment (lildonpancho) [inline README.md:68]: We will make the scripts executable using something like `chmod +x run.sh`. It won't be necessary to run them using the bash command.

**Review by lildonpancho — CHANGES_REQUESTED** [review]
Sorry for the minor tweaks. Just trying to make sure things are good to go for everyone as we move forward. Also, did you ever get a notice that Walgreens had openings when you ran the Docker container? I think by containerizing it, we inadvertently fixed the Walgreens bug.

File: scripts/run.sh (hunk 5 above)
Comment (cassjs) [reply scripts/run.sh]: I removed the -i per our conversation on slack. ✅

File: README.md (hunk 6 above)
Comment (cassjs) [reply README.md]: Done and edited the readme with new commands. ✅

**Comment by cassjs** [comment]
> Sorry for the minor tweaks. Just trying to make sure things are good to go for everyone as we move forward. Also, did you ever get a notice that Walgreens had openings when you ran the Docker container? I think by containerizing it, we inadvertently fixed the Walgreens bug.

No worries, this feedback has been so valuable to me! I'm learning a lot! 

Walgreens is now working. ✨ It's a Docker miracle!!!!✨ 
Screen Shot 2021-04-27 at 1 12 02 PM

**Review by lildonpancho — APPROVED** [review]
Everything is working. Great job!

**Comment by cassjs** [comment]
> Everything is working. Great job!

Thanks for your help!

**Comment by Xuehong-pdx** [comment]
I got an error " image " when try to login in from the vaccinewatcher folder.  Docker ID and password were correct since i was able to login to the desktop version. Not sure what caused this.

**Comment by Xuehong-pdx** [comment]
Good job.  I was able to build and run docker files.  The only thing I noticed that when I tried to run  "$ docker login", I got an error message I think it is because I already logged in on my desktop.  It should prompted "you are already login" or something like that.  This seems to be a problem on docker's end.  I ignored this error message and was able to run ''$ ./scripts/build.sh" and "$ ./scripts/build.sh" successfully.  Before docker, I was able to use the CLI() options to get data for Portland, but cannot do that any more.

**Comment by lildonpancho** [comment]
@Xuehong-pdx that has to do with the fact that we are not accepting command-line arguments for the run.sh script to feed into the Docker run command. There is a little bash scripting that could fix that. @cassjs, or you could work on that. It shouldn't take too long.

**Comment by street-jackal** [comment]
> > Sorry for the minor tweaks. Just trying to make sure things are good to go for everyone as we move forward. Also, did you ever get a notice that Walgreens had openings when you ran the Docker container? I think by containerizing it, we inadvertently fixed the Walgreens bug.
> 
> No worries, this feedback has been so valuable to me! I'm learning a lot!
> 
> Walgreens is now working. ✨ It's a Docker miracle!!!!✨
> Screen Shot 2021-04-27 at 1 12 02 PM

**Comment by cassjs** [comment]
I have some free time today and can give it a try.

**Comment by Xuehong-pdx** [comment]
> 
> 
> @Xuehong-pdx that has to do with the fact that we are not accepting command-line arguments for the run.sh script to feed into the Docker run command. There is a little bash scripting that could fix that. @cassjs, or you could work on that. It shouldn't take too long.
> 
> Example:
> 
> ```
> usage="Usage: $0 <VERSION TAG> <FLASK_ENV>"
> 
> if [ "$1" ]
> then
>   STATE = $1
> else
>   STATE = "Texas"
> fi
> 
> docker run vaccine-watcher --state=STATE
> ```
> 
> Or we can just set some env variables in the script and feed them in as you see fit to test different things. Up to you. The second is easier the first is fancier. Totally up to you.

Do you have a good source for me to read about this? Not sure about the syntax for " usage="Usage: $0 <VERSION TAG> <FLASK_ENV>"etc.  Haven't seen this in the docker doc yet.  Thanks

**Comment by lildonpancho** [comment]
@Xuehong-pdx sorry copy and paste error. I fixed it.

**Comment by Xuehong-pdx** [comment]
> 
> 
> > @Xuehong-pdx that has to do with the fact that we are not accepting command-line arguments for the run.sh script to feed into the Docker run command. There is a little bash scripting that could fix that. @cassjs, or you could work on that. It shouldn't take too long.
> > Example:
> > ```
> > if [ "$1" ]
> > then
> >   STATE = $1
> > else
> >   STATE = "Texas"
> > fi
> > 
> > docker run vaccine-watcher --state=STATE
> > ```
> > 
> > 
> > Or we can just set some env variables in the script and feed them in as you see fit to test different things. Up to you. The second is easier the first is fancier. Totally up to you.
> 
> what does "if [$1] " check?   I assumed it was checking the input, but when I included the above in run.sh and ran it with or without --state= 'Texas" . I first get a STATE not found error before the logging message.  Then within the message it assigned state=state, rather than 'Texas" as the else statement specified above.  Thanks

**Comment by street-jackal** [comment]
So I got Docker setup per the readme (thank you!), but I'm getting an error when I call run.sh:

`bash-3.2$ ./scripts/run.sh
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested`

Seems like this is related to my shiny ARM Macbook. Is there a way to add M1/ARM compatibility to the docker image? 

More here: [https://docs.docker.com/docker-for-mac/apple-silicon/](url)

**Comment by lildonpancho** [comment]
@street-jackal Looks like this is just a warning. Does the Docker container quit after you receive this message?

**Comment by street-jackal** [comment]
Yes it does, sorry I should have posted the traceback:
`WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
Traceback (most recent call last):
  File "watcher.py", line 313, in <module>
    cli()
  File "watcher.py", line 295, in cli
    watcher = get_vaccine_watcher(config=params, freq_secs=args.freq, hook=hook, check_walgreens=args.wg, check_cvs=args.cvs, verbose=args.verbose)
  File "watcher.py", line 250, in get_vaccine_watcher
    configure_watcher(**config)
  File "watcher.py", line 246, in configure_watcher
    _watcher = VaccineWatcher(**config)
  File "watcher.py", line 118, in __init__
    self.browser = self.api.browser
  File "watcher.py", line 79, in browser
    self._create_browser()
  File "watcher.py", line 67, in _create_browser
    self._create_driver()
  File "watcher.py", line 61, in _create_driver
    self._driver = webdriver.Chrome(options=self.chrome_options, seleniumwire_options=self.selenium_wire_options, executable_path=self.exec_path)
  File "/home/seluser/.local/lib/python3.8/site-packages/undetected_chromedriver/__init__.py", line 53, in __new__
    instance.__init__(*args, **kwargs)
  File "/home/seluser/.local/lib/python3.8/site-packages/seleniumwire/webdriver.py", line 114, in __init__
    super().__init__(*args, **kwargs)
  File "/home/seluser/.local/lib/python3.8/site-packages/selenium/webdriver/chrome/webdriver.py", line 76, in __init__
    RemoteWebDriver.__init__(
  File "/home/seluser/.local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 157, in __init__
    self.start_session(capabilities, browser_profile)
  File "/home/seluser/.local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 252, in start_session
    response = self.execute(Command.NEW_SESSION, parameters)
  File "/home/seluser/.local/lib/python3.8/site-packages/selenium/webdriver/remote/webdriver.py", line 321, in execute
    self.error_handler.check_response(response)
  File "/home/seluser/.local/lib/python3.8/site-packages/selenium/webdriver/remote/errorhandler.py", line 242, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.WebDriverException: Message: unknown error: Chrome failed to start: crashed.
  (unknown error: DevToolsActivePort file doesn't exist)
  (The process started from chrome location /usr/bin/google-chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.)`

**Comment by lildonpancho** [comment]
Can you edit the script with `docker run --platform linux/arm64/v8 vaccine-watcher`?

**Comment by cassjs** [comment]
I'm stuck @lildonpancho. I'm trying to run the run.sh script but it's saying the commands are not found. I haven't written many bash scripts. Am I missing anything? 

Screen Shot 2021-04-27 at 7 33 35 PM

**Pull request merged**